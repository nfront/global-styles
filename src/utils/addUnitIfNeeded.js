import unitless from './unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function addUnitIfNeeded(name, value) {
  if (value == null || typeof value === 'boolean' || value === '') {
    return '';
  }

  if (typeof value === 'number' && value !== 0 && !(name in unitless)) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
