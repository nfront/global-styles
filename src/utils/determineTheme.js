export default (props, fallbackTheme) => {
  // Props should take precedence over ThemeProvider, which should take precedence over
  // defaultProps, but React automatically puts defaultProps on props.

  const theme = props.theme ? props.theme : fallbackTheme;

  return theme;
};
