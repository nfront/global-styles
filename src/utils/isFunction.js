export default function isFunction(test) {
  return typeof test === 'function';
}
