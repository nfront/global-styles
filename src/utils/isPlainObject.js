export default x => typeof x === 'object' && x.constructor === Object;
