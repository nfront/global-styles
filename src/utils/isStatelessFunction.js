export default function isStatelessFunction(test) {
  return typeof test === 'function' && !(test.prototype && test.prototype.isReactComponent);
}
