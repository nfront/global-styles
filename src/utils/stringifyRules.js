const COMMENT_REGEX = /^\s*\/\/.*$/gm;

export default function stringifyRules(rules, selector, prefix) {
  const flatCSS = rules.join('').replace(COMMENT_REGEX, ''); // replace JS comments

  const cssStr = selector && prefix ? `${prefix} ${selector} { ${flatCSS} }` : flatCSS;

  return cssStr;
}
