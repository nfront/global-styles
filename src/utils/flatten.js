import isFunction from './isFunction';
import isStatelessFunction from './isStatelessFunction';
import isPlainObject from './isPlainObject';
import hyphenate from './hyphenateStyleName';
import addUnitIfNeeded from './addUnitIfNeeded';

/**
 * It's falsish not falsy because 0 is allowed.
 */
const isFalsish = chunk => {
  return chunk === undefined || chunk === null || chunk === false || chunk === '';
};

const objToCss = (obj, prevKey) => {
  const css = Object.keys(obj)
    .filter(key => !isFalsish(obj[key]))
    .map(key =>
      isPlainObject(obj[key]) ? objToCss(obj[key], key) : `${hyphenate(key)}: ${addUnitIfNeeded(key, obj[key])};`
    )
    .join(' ');
  return prevKey
    ? `${prevKey} {
  ${css}
}`
    : css;
};

export default function flatten(chunk, executionContext) {
  if (Array.isArray(chunk)) {
    const ruleSet = [];

    for (let i = 0, len = chunk.length, result; i < len; i += 1) {
      result = flatten(chunk[i], executionContext);

      if (result != null) {
        if (Array.isArray(result)) ruleSet.push(...result);
        else ruleSet.push(result);
      }
    }

    return ruleSet;
  }

  if (isFalsish(chunk)) {
    return null;
  }

  /* Either execute or defer the function */
  if (isFunction(chunk)) {
    if (isStatelessFunction(chunk) && executionContext) {
      const result = chunk(executionContext);

      return flatten(result, executionContext);
    }
    return chunk;
  }

  /* Handle objects */
  return isPlainObject(chunk) ? objToCss(chunk) : chunk.toString();
}
