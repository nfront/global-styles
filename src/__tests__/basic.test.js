/**
 * @jest-environment jsdom
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import { createGlobalStyle, css } from '../index';
import { expectCSSMatches, stripCWSArray, stripCWS } from './utils';
import flatten from '../utils/flatten';
import hashCode from '../utils/hashCode';
import isStatelessFunction from '../utils/isStatelessFunction';
import stringifyRules from '../utils/stringifyRules';

describe('css()', () => {
  test('Without interpolations, css() just returns array with css string as its only element', () => {
    const cssResult = stripCWSArray(css`
      .div {
        color: blue;
      }
    `);
    expect(cssResult).toEqual([`.div { color: blue; }`]);
  });

  test('Without arguments passed to css(), it just returns interleaved CSS strings and untouched functions', () => {
    const cssResult = stripCWSArray(css`
      body {
        color: ${props => (props.light ? 'white' : 'black')};
        font-family: ${props => props.theme.typography.fontFamily};
      }
    `);

    const expectation = stripCWSArray([
      `body {
      color: `,
      props => (props.light ? 'white' : 'black'),
      `;
      font-family: `,
      props => props.theme.typography.fontFamily,
      `;
    }`,
    ]);

    expect(cssResult).toEqual(expectation);
  });

  test('css() can convert plain JS objects to css strings', () => {
    expect(stripCWSArray(css({ html: { marginTop: '15px' } }))).toEqual(['html { margin-top: 15px; }']);
  });
});

describe('flatten', () => {
  test('flatten adds px units to objects that require it, when converting to css string', () => {
    expect(stripCWS(flatten({ html: { marginTop: 15 } }))).toEqual('html { margin-top: 15px; }');
  });

  test('Can convert empty object to empty css string', () => {
    expect(flatten({})).toEqual('');
  });

  test('Can use boolen as object value', () => {
    expect(stripCWS(flatten({ html: { marginTop: true } }))).toEqual('html { margin-top: ; }');
  });
});

describe('GlobalStyleComponent', () => {
  let reset;
  let globalStyle;
  let GlobalStyleComponent;
  let outputCss;

  beforeAll(() => {
    reset = css`
      .div {
        color: blue;
      }
    `;

    globalStyle = css`
      .my-class2 {
        margin-bottom: 10rem;
        -webkit-border-radius: 4px;
        -moz-border-radius: 4px;
        border-radius: 4px;
      }
      html {
        background-color: blue;
      }
      body {
        color: ${props => (props.light ? 'white' : 'black')};
        font-family: ${props => props.theme.typography.fontFamily};
      }
    `;

    GlobalStyleComponent = createGlobalStyle`${reset}${globalStyle}`;
  });

  test('createGlobalStyle() can accept empty ruleset, and will then not inject anything in head.', () => {
    createGlobalStyle``;
    expectCSSMatches('');
  });

  test('GlobalStyleComponent inserts after another head style, if one exists', () => {
    const node = document.createElement('style');
    node.id = 'otherStyle';
    node.innerHTML = 'html { color: blue; }';
    const { head } = document;
    head.appendChild(node);

    const component = TestRenderer.create(
      <GlobalStyleComponent dark theme={{ typography: { fontFamily: 'Roboto' } }} />
    );

    const styleArray = document.querySelectorAll('style');
    expect(styleArray.length).toBe(2);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    styleArray.forEach(style => style.parentElement.removeChild(style));
  });

  test('GlobalStyleComponent injects correct CSS styles in head', () => {
    const component = TestRenderer.create(
      <GlobalStyleComponent light theme={{ typography: { fontFamily: 'Arial' } }} />
    );

    outputCss = `.div {
      color: blue;
    }
    .my-class2 {
      margin-bottom: 10rem;
      -webkit-border-radius: 4px;
      -moz-border-radius: 4px;
      border-radius: 4px;
    }
    html {
      background-color: blue;
    }
    body {
      color: white;
      font-family: Arial;
    }`;

    expectCSSMatches(outputCss);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('GlobalStyleComponent overrides old head style, if style already exists', () => {
    const component = TestRenderer.create(
      <GlobalStyleComponent dark theme={{ typography: { fontFamily: 'Roboto' } }} />
    );

    outputCss = `.div {
      color: blue;
    }
    .my-class2 {
      margin-bottom: 10rem;
      -webkit-border-radius: 4px;
      -moz-border-radius: 4px;
      border-radius: 4px;
    }
    html {
      background-color: blue;
    }
    body {
      color: black;
      font-family: Roboto;
    }`;

    expectCSSMatches(outputCss);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('GlobalStyleComponent with default theme', () => {
    const component = TestRenderer.create(<GlobalStyleComponent dark />);

    outputCss = `.div {
      color: blue;
    }
    .my-class2 {
      margin-bottom: 10rem;
      -webkit-border-radius: 4px;
      -moz-border-radius: 4px;
      border-radius: 4px;
    }
    html {
      background-color: blue;
    }
    body {
      color: black;
      font-family: Arial;
    }`;

    expectCSSMatches(outputCss);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('GlobalStyleComponent can return a React component, which first runs createStyles(props), but does not inject anything into head.', () => {
    const { ReactStyleComponent } = GlobalStyleComponent.globalStyle;
    const component = TestRenderer.create(
      <ReactStyleComponent light theme={{ typography: { fontFamily: 'Arial' } }} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('GlobalStyleComponent should give warning if run with children', () => {
    const component = TestRenderer.create(<GlobalStyleComponent>child</GlobalStyleComponent>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('other small functions', () => {
  test('hashCode if 0 length string', () => {
    expect(hashCode('')).toBe(0);
  });

  test('isStatelessFunction with normal function', () => {
    expect(isStatelessFunction(() => {})).toBe(true);
  });

  test('isStatelessFunction with React function', () => {
    // eslint-disable-next-line react/prefer-stateless-function
    class ReactComp extends React.Component {
      render() {
        return <h1>Hello</h1>;
      }
    }
    expect(isStatelessFunction(ReactComp)).toBe(false);
  });

  test('stringifyRules can be called with selector and prefix specified seperately.', () => {
    expect(() => stringifyRules(['div { color: red; }'], 'html', ':not')).not.toThrow();
  });
});
