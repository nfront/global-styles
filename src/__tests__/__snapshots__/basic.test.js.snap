// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalStyleComponent GlobalStyleComponent can return a React component, which first runs createStyles(props), but does not inject anything into head. 1`] = `
<style
  dangerouslySetInnerHTML={
    Object {
      "__html": "
      .div {
        color: blue;
      }
    
      .my-class2 {
        margin-bottom: 10rem;
        -webkit-border-radius: 4px;
        -moz-border-radius: 4px;
        border-radius: 4px;
      }
      html {
        background-color: blue;
      }
      body {
        color: white;
        font-family: Arial;
      }
    ",
    }
  }
  id="GlobalStyle404084579"
/>
`;

exports[`GlobalStyleComponent GlobalStyleComponent injects correct CSS styles in head 1`] = `null`;

exports[`GlobalStyleComponent GlobalStyleComponent inserts after another head style, if one exists 1`] = `null`;

exports[`GlobalStyleComponent GlobalStyleComponent overrides old head style, if style already exists 1`] = `null`;

exports[`GlobalStyleComponent GlobalStyleComponent should give warning if run with children 1`] = `null`;

exports[`GlobalStyleComponent GlobalStyleComponent with default theme 1`] = `null`;
